When the attachment "progress" is zero, show a placeholder icon

squelart opened the pull request:
#### 1f2261656aa215fbf8d4acde1d2986a6f205c5ce
<pre>
When the attachment &quot;progress&quot; is zero, show a placeholder icon
<a href="https://bugs.webkit.org/show_bug.cgi?id=256292">https://bugs.webkit.org/show_bug.cgi?id=256292</a>
rdar://108266781

Reviewed by Aditya Keerthi and Tim Horton.

The arrow-down-circle placeholder is used to show that the attachment still needs to be downloaded.

* LayoutTests/fast/attachment/cocoa/wide-attachment-rendering.html:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentPlaceholderIdentifier):
(WebCore::attachmentProgressCSSProperty):
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::updateProgress):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-placeholder):
(div#attachment-save-icon):

Canonical link: <a href="https://commits.webkit.org/263696@main">https://commits.webkit.org/263696@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/09f795a7515a56bf4b3f93adf5f1e5c91772e316

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5398 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5537 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/5727 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/6939 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/5427 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5397 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/5765 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5517 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/6333 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5499 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5565 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/4832 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/6968 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3044 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/4839 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/11995 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/4910 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/4918 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/6591 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5353 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4395 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/4813 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/4836 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1306 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/8914 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5175 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -291,10 +299,17 @@ void HTMLAttachmentElement::updateProgress(const AtomString& progress)
     bool validProgress = false;
     float value = progress.toFloat(&validProgress);
     if (validProgress && std::isfinite(value)) {
+        if (value == 0.0) {
```
Do we want a `areEssentiallyEqual` here or is it really really 0?

squelart replied:
I think we really want 0.0, as we just want to check for `progress="0"`, and that matches the check in the legacy attachment code.

pxlcoder replied:
Can we just do `!value` then?

squelart replied:
I personally hate that for numbers, but fiiine, if that's the preferred style... 😬

pxlcoder replied:
It is 😅 

https://webkit.org/code-style-guidelines/#zero-comparison

hortont424 commented on Source/WebCore/html/HTMLAttachmentElement.cpp:
```diff
@@ -291,10 +299,17 @@ void HTMLAttachmentElement::updateProgress(const AtomString& progress)
     if (validProgress && std::isfinite(value)) {
+        if (value == 0.0) {
+            m_placeholderElement->removeAttribute(styleAttr);
+            m_progressElement->setAttributeWithoutSynchronization(styleAttr, "display: none;"_s);
```
Vaguely surprised we're setting `style` like this instead of `setInlineStyleProperty` like PDFPlugin&friends; isn't it hard to keep track/avoid overwriting things you set from elsewhere?

squelart replied:
I'll look into `setInlineStyleProperty`, nifty!

pxlcoder replied:
`setInlineStyleProperty` would be good. I think you'll have to keep the `--progress:` below using the attribute setter though.

squelart replied:
`setInlineStyleCustomProperty` seems to work for `--progress`.

pxlcoder replied:
Nice!

hortont424 commented on Source/WebCore/html/shadow/attachmentElementShadow.css:
```diff
@@ -81,6 +81,18 @@ attachment#attachment-preview {
+    grid-column: 1;
+    width: 100%;
+    aspect-ratio: 1;
+    /* SF Symbol arrow.down.circle */
```
Can you, when importing symbols, specify the size and weight too in your comment?

squelart replied:
I got them from https://copysfsymbols.apple.com/index.html, no size options there! I'll add a note...

pxlcoder replied:
We can probably say they are "Medium", since that's the default.

squelart committed:
When the attachment "progress" is zero, show a placeholder icon
https://bugs.webkit.org/show_bug.cgi?id=256292
rdar://108266781

Reviewed by Aditya Keerthi and Tim Horton.

The arrow-down-circle placeholder is used to show that the attachment still needs to be downloaded.

* LayoutTests/fast/attachment/cocoa/wide-attachment-rendering.html:
* LayoutTests/platform/ios-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* LayoutTests/platform/mac-wk2/fast/attachment/cocoa/wide-attachment-rendering-expected.txt:
* Source/WebCore/html/HTMLAttachmentElement.cpp:
(WebCore::attachmentPlaceholderIdentifier):
(WebCore::attachmentProgressCSSProperty):
(WebCore::HTMLAttachmentElement::ensureModernShadowTree):
(WebCore::HTMLAttachmentElement::updateProgress):
* Source/WebCore/html/HTMLAttachmentElement.h:
* Source/WebCore/html/shadow/attachmentElementShadow.css:
(div#attachment-placeholder):
(div#attachment-save-icon):

Canonical link: https://commits.webkit.org/263696@main
